### PR TITLE
Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by ramiaquista**
## What we have done 🔨  :
- Create a Book class in a separate .rb file.
- Create a Label class with an association to the Item class (in a separate .rb file).
- All Book class properties visible in the diagram should be defined and set up in the constructor method.
- All Label class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
   `add_item` method in the Label class
    should take an instance of the Item class as an input
    should add the input item to the collection of items
    should add self as a property of the item object (by using the correct setter from the item object)
  ` can_be_archived?()` in the Book class
    should override the method from the parent class
    should return true if parent's method returns true OR if cover_state equals to "bad"
    otherwise, it should return false
- Add unit tests for all implemented methods.
- The following options should be available:
   List all books
   List all labels (e.g. 'Gift', 'New')
   Add a book
- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
   books table (add all properties and associations from the parent Item class as table columns)
   labels table
- Created MusicAlbum class in a separate .rb file.
- Created Genre class with an association to the Item class (in a separate .rb file).
- defined All MusicAlbum class properties visible in the diagram and set up in the constructor method.
- defined All Genre class properties visible in the diagram and set up in the constructor method.
- added add_item method in the Genre class
- Added can_be_archived?() in the MusicAlbum class
- Added unit tests for all implemented methods.
- Listed all music albums
- Listed all genres (e.g 'Comedy', 'Thriller')
- Added a music album
- Implemented All data saving in .json files.
- Created a schema.sql file with tables that will be analogical to the structure of the:
      1. music_albums table (add all properties and associations from the parent Item class as table columns)
      2. genres table
- Created `game` class inside `game.rb` file.
- Added `author` class inside `author.rb` file.
- Added `can_be_archived?()` method to Game class.
- Added random id to author class.
- Added unit tests for `Game` class and methods.
- Added unit tests for `Author` class and methods.
- Added logic to list options.
- Added logic files to read and store data in json files.
- Added logic to store and read authors json data.
- Added `schema.sql` with item, game and author tables

### Project Description video 🎥  : 
- [Link to the video](https://drive.google.com/file/d/15BlWCxXdBryXtuoBWou-g5WLIOZutzXR/view?usp=sharing)

### CODE REVIEW

File: item.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @author = author
+  end
+
+  def add_source(source)
+    @source = source
+  end
+
+  def add_label(label)
+    @label = label
+  end
```
Comment (Lameck1) [inline item.rb:26]: - [ ] Please the above suggestion as a guideline to refactor these methods in a bid to properly implement the associations.

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    list_all_labels
+  end
+end
+
+def second_conditions(num)
+  case num
+  when 7
+    list_all_authors
+  when 8
+    list_all_sources
```
Comment (Lameck1) [inline main.rb:48]: - In the App, as tested, when the user chooses option `8`, the following bug 👇🏼  is encountered. Please refactor accordingly.

image

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  when 9
+    add_book
+  when 10
+    add_music_album
+  when 11
+    add_movie
```
Comment (Lameck1) [inline main.rb:54]: - In the App, as tested, when the user chooses option `11`, the following bug 👇🏼  is encountered. Please refactor accordingly.


image

File: __tests__/music_album_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+require './music_album'
+
+describe Music_alboum do
+  context 'Creating an instance of a musicAlbum' do
+    it 'should create a new instace of a musicAlbum' do
+      music_album = MusicAlbum.new('22/02/2022')
+      expect(music_album.on_spotify).to eq(false)
+    end
```
Comment (Lameck1) [inline __tests__/music_album_spec.rb:8]: - [ ] Once the above error is corrected, the tests will run but this particular `it` block will still be failing as highlighted below. Please take time and refactor.

image

**Review by Lameck1 — CHANGES_REQUESTED** [review]
Hi, @ramiaquista, @mustabbas, & @elassis, 

Good job so far!

### Some highlights
* The team followed correct gitflow practice ✔️ 
* There is evidence of proper collaboration ✔️ 
* Good commit messages ✔️ 
* Proffessional README ✔️ 
* Awesome video presentation ✔️ 

However, we still have some issues to iron out before we can move ahead.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: __tests__/music_album_spec.rb (hunk 4 above)
Comment (ramiaquista) [reply __tests__/music_album_spec.rb:8]: Hi @Lameck1, thank's for the amazing review, it was very descriptive and easy to implement! 😄 

We have implemented the requested changes, so let's hope it's better now! 🔥

Comment (Lameck1) [reply __tests__/music_album_spec.rb:8]: You are welcome @ramiaquista. Thanks.

**Review by Lameck1 — APPROVED** [review]
Hi, @ramiaquista, @mustabbas, & @elassis,

You've done well addressing the concerns from the previous review 👏🏼.
Your project is now complete!  Consider merging this branch :shipit: 🚀 

Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**